src/Domain/Item/ItemService.ts

```diff
@@ -49,8 +55,138 @@ export class ItemService implements ItemServiceInterface {
     }
   }
 
-  async saveItems(_itemHashes: string[], _userAgent: string, _retrievedItems: Item[]): Promise<SaveItemsResult> {
-    throw new Error('Method not implemented.')
+  async saveItems(dto: SaveItemsDTO): Promise<SaveItemsResult> {
+    const savedItems: Array<Item> = []
+    const conflicts: Array<ItemConflict> = []
+
+    const lastUpdatedTimestamp = this.timer.getTimestampInMicroseconds()
+
+    for (const itemHash of dto.itemHashes) {
+      const existingItem = await this.itemRepository.findByUuidAndUserUuid(itemHash.uuid, dto.userUuid)
+
+      if (!this.itemShouldBeSaved(itemHash, existingItem)) {
+        conflicts.push({
+          serverItem: existingItem,
+          type: 'sync_conflict'
+        })
+
+        continue
+      }
+
+      if(existingItem) {
+        const updatedItem = await this.updateExistingItem(existingItem, itemHash, dto.userAgent)
+        savedItems.push(updatedItem)
+      } else {
+        try {
+          const newItem = await this.saveNewItem(itemHash, dto.userAgent)
+          savedItems.push(newItem)
+        } catch (error) {
+          conflicts.push({
+            unsavedItem: itemHash,
+            type: 'uuid_conflict',
+          })
+
+          continue
+        }
+      }
+    }
+
+    const syncToken = this.calculateSyncToken(lastUpdatedTimestamp, savedItems)
+
+    return {
+      savedItems,
+      conflicts,
+      syncToken
+    }
+  }
+
+  async frontLoadKeysItemsToTop(userUuid: string, retrievedItems: Array<Item>): Promise<Array<Item>> {
+    const itemsKeys = await this.itemRepository.findAll({
```

Comment:
Hmm..I'm not following why we would need to fetch all items keys from the db initially. Why can't we just filter the incoming `retrievedItems` by `contentType` and move them to the top?

Reply:
I think it's a 1-to-1 state with what we have in Ruby: https://github.com/standardnotes/syncing-server/blob/develop/lib/sync_engine/2020_01_15/sync_manager.rb#L18-L20  where `@user.items_keys` is essentially `items.where(content_type: 'SN|ItemsKey')`.

Not sure what are the reasons behind this - maybe it has to do with the `limit` so maybe a scenario where the itemskey is not among the loaded items due to the limit?

Reply:
Ah right I see 👍 Seemingly on my end this should have just been filtering the retrieved_items array without having to query the user's items key. Seems to logically be the same thing.